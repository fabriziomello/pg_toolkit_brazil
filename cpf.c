#include "postgres.h"
#include "fmgr.h"
#include "utils/fmgrprotos.h"
#include "utils/palloc.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(cpfout);

Datum
cpfout(PG_FUNCTION_ARGS)
{
	int64		val = PG_GETARG_INT64(0);
	int			group[4];
	char	   *result;

	/* 000.000.000-00 */
	group[0] = (val / 100000000) % 1000;	/* first group of three digits */
	group[1] = (val / 100000) % 1000;	/* second group of three digits */
	group[2] = (val / 100) % 1000;	/* third group of three digits */
	group[3] = (val % 100);		/* check digits (last two digits) */

	result = psprintf("%03d.%03d.%03d-%02d", group[0], group[1], group[2], group[3]);

	PG_RETURN_CSTRING(result);
}

/* https://pt.wikipedia.org/wiki/Cadastro_de_Pessoas_F%C3%ADsicas */
static int
compute_cpf_check_digits(long int cpf)
{
	int			i = 0,
				d = 0,
				d1 = 0,
				d2 = 0;

	/* remove last two digits */
	cpf = cpf / 100;

	do
	{
		d = cpf % 10;
		d1 = d1 + d * (9 - (i % 10));
		d2 = d2 + d * (9 - ((i++ + 1) % 10));
		cpf = cpf / 10;
	} while (cpf > 0);

	d1 = (d1 % 11) % 10;
	d2 = d2 + d1 * 9;
	d2 = (d2 % 11) % 10;

	return (d1 * 10) + d2;
}

static void
validate_cpf(int64 cpf)
{
	char	   *out;
	int			i;

	/* Check sizes */
	if (cpf < 100 || cpf > 99999999999)
	{
		ereport(ERROR,
				(errcode(ERRCODE_WRONG_OBJECT_TYPE),
				 errmsg("invalid CPF"),
				 errdetail("CPF should not be less than 100 and greater than 99999999999.")));
	}

	/* Check if all digits are the same. */
	out = psprintf("%011ld", cpf);
	for (i = 1; i < 11 && out[i] == out[0]; i++)
	{
		if (i == 10)
		{
			ereport(ERROR,
					(errcode(ERRCODE_WRONG_OBJECT_TYPE),
					 errmsg("invalid CPF"),
					 errdetail("All CPF digits should not be equal.")));
		}
	}

	if (compute_cpf_check_digits(cpf) != (cpf % 100))
	{
		ereport(ERROR,
				(errcode(ERRCODE_WRONG_OBJECT_TYPE),
				 errmsg("invalid CPF"),
				 errdetail("Invalid check digit for the CPF.")));
	}
}

PG_FUNCTION_INFO_V1(cpfin);

Datum
cpfin(PG_FUNCTION_ARGS)
{
	int64		value = DirectFunctionCall1(int8in, PG_GETARG_DATUM(0));

	validate_cpf(value);

	PG_RETURN_INT64(value);
}

PG_FUNCTION_INFO_V1(cpfbigint);

Datum
cpfbigint(PG_FUNCTION_ARGS)
{
	int64		value = PG_GETARG_INT64(0);

	validate_cpf(value);

	PG_RETURN_INT64(value);
}
