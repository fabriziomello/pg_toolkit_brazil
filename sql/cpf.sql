CREATE EXTENSION pg_toolkit_brazil;

CREATE TABLE pessoa(
    cpf cpf NOT NULL,
    PRIMARY KEY (cpf)
);
\d pessoa
-- Valid and generated CPFs using: https://theonegenerator.com/pt/geradores/documentos/gerador-de-cpf/
INSERT INTO pessoa VALUES (cpf '40100276300'), (16426332836);
SELECT * FROM pessoa ORDER BY cpf;

-- invalid CPFs
SELECT cpf '12345';
SELECT cpf '99';
SELECT cpf '123456789012';
SELECT cpf '11111111111';
SELECT cpf '22222222222';
SELECT cpf '33333333333';
SELECT cpf '44444444444';
SELECT cpf '55555555555';
SELECT cpf '66666666666';
SELECT cpf '77777777777';
SELECT cpf '88888888888';
SELECT cpf '99999999999';
SELECT cpf '00000000000';

-- operators
SELECT cpf '40100276300' =  cpf '40100276300';
SELECT cpf '40100276300' <> cpf '16426332836';
SELECT cpf '40100276300' <  cpf '16426332836';
SELECT cpf '40100276300' <= cpf '16426332836';
SELECT cpf '40100276300' >= cpf '16426332836';
SELECT cpf '40100276300' >  cpf '16426332836';