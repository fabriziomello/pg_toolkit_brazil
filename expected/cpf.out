CREATE EXTENSION pg_toolkit_brazil;
CREATE TABLE pessoa(
    cpf cpf NOT NULL,
    PRIMARY KEY (cpf)
);
\d pessoa
             Table "public.pessoa"
 Column | Type | Collation | Nullable | Default 
--------+------+-----------+----------+---------
 cpf    | cpf  |           | not null | 
Indexes:
    "pessoa_pkey" PRIMARY KEY, btree (cpf)

-- Valid and generated CPFs using: https://theonegenerator.com/pt/geradores/documentos/gerador-de-cpf/
INSERT INTO pessoa VALUES (cpf '40100276300'), (16426332836);
SELECT * FROM pessoa ORDER BY cpf;
      cpf       
----------------
 164.263.328-36
 401.002.763-00
(2 rows)

-- invalid CPFs
SELECT cpf '12345';
ERROR:  invalid CPF
LINE 1: SELECT cpf '12345';
                   ^
DETAIL:  Invalid check digit for the CPF.
SELECT cpf '99';
ERROR:  invalid CPF
LINE 1: SELECT cpf '99';
                   ^
DETAIL:  CPF should not be less than 100 and greater than 99999999999.
SELECT cpf '123456789012';
ERROR:  invalid CPF
LINE 1: SELECT cpf '123456789012';
                   ^
DETAIL:  CPF should not be less than 100 and greater than 99999999999.
SELECT cpf '11111111111';
ERROR:  invalid CPF
LINE 1: SELECT cpf '11111111111';
                   ^
DETAIL:  All CPF digits should not be equal.
SELECT cpf '22222222222';
ERROR:  invalid CPF
LINE 1: SELECT cpf '22222222222';
                   ^
DETAIL:  All CPF digits should not be equal.
SELECT cpf '33333333333';
ERROR:  invalid CPF
LINE 1: SELECT cpf '33333333333';
                   ^
DETAIL:  All CPF digits should not be equal.
SELECT cpf '44444444444';
ERROR:  invalid CPF
LINE 1: SELECT cpf '44444444444';
                   ^
DETAIL:  All CPF digits should not be equal.
SELECT cpf '55555555555';
ERROR:  invalid CPF
LINE 1: SELECT cpf '55555555555';
                   ^
DETAIL:  All CPF digits should not be equal.
SELECT cpf '66666666666';
ERROR:  invalid CPF
LINE 1: SELECT cpf '66666666666';
                   ^
DETAIL:  All CPF digits should not be equal.
SELECT cpf '77777777777';
ERROR:  invalid CPF
LINE 1: SELECT cpf '77777777777';
                   ^
DETAIL:  All CPF digits should not be equal.
SELECT cpf '88888888888';
ERROR:  invalid CPF
LINE 1: SELECT cpf '88888888888';
                   ^
DETAIL:  All CPF digits should not be equal.
SELECT cpf '99999999999';
ERROR:  invalid CPF
LINE 1: SELECT cpf '99999999999';
                   ^
DETAIL:  All CPF digits should not be equal.
SELECT cpf '00000000000';
ERROR:  invalid CPF
LINE 1: SELECT cpf '00000000000';
                   ^
DETAIL:  CPF should not be less than 100 and greater than 99999999999.
-- operators
SELECT cpf '40100276300' =  cpf '40100276300';
 ?column? 
----------
 t
(1 row)

SELECT cpf '40100276300' <> cpf '16426332836';
 ?column? 
----------
 t
(1 row)

SELECT cpf '40100276300' <  cpf '16426332836';
 ?column? 
----------
 f
(1 row)

SELECT cpf '40100276300' <= cpf '16426332836';
 ?column? 
----------
 f
(1 row)

SELECT cpf '40100276300' >= cpf '16426332836';
 ?column? 
----------
 t
(1 row)

SELECT cpf '40100276300' >  cpf '16426332836';
 ?column? 
----------
 t
(1 row)

